Switch from github.com/form3tech-oss/jwt-go to github.com/golang-jwt/jwt/v4

jlourenc opened the pull request:
Fixes #355 
Deprecates #370 
Follows-up on #344

jlourenc commented:
Hi @crewjam, would you have time to review this PR?

crewjam reviewed (approved):
LGTM, thanks! :)

jlourenc commented:
Go 1.17 seems to be responsible for the failing tests. Running the tests on **main** also fail. They successfully pass using Go 1.16.

crewjam commented:
Yes, I agree. I'd like to get that sorted out before we merge anything else. :/

jlourenc commented:
I believe I've tracked down the changes of behaviour in Go 1.17, introduced by:
1. https://go-review.googlesource.com/c/go/+/277892
2. https://go-review.googlesource.com/c/go/+/277893

This causes https://github.com/mattermost/xml-roundtrip-validator to no longer detect ill-formed XML since leading, trailing or multiple colons in attributes would be assigned to its name.

jlourenc commented:
Hey @crewjam , have you had time to look into the Go 1.17 change of behaviour?

urgas9 commented:
@crewjam any chance we merge this and release a new version of the lib? I am looking for the dependency fix ☝️ Thanks @jlourenc for the fix

SVilgelm commented:
I still don't understand why, but the github.com/form3tech-oss/jwt-go leads in having the github.com/dgrijalva/jwt-go as indirect dependency and it blocks us in upgrading other deps

SVilgelm commented:
so, @crewjam would you be able to merge this PR and bump new release? seems the go 17 issue is unrelated to the changes made in this PR

crewjam commented:
Thanks for your patience on this. I’m not dead just super busy with day job, but yes I think we can merge. I’d like to get the test passing/stable too. :)

beorc commented:
@crewjam Regarding the failing test: https://github.com/crewjam/saml/blob/main/samlsp/fetch_metadata_test.go#L31

Looks like it actually should fail using Go 1.17 :
https://github.com/mattermost/xml-roundtrip-validator/commit/460296229913fb6653021a48599a3a0ff83802e6#diff-a0bfeed7cbb7b92c[…]722ba64e9e777ab1f8cR28

And should not using Go 1.16:
https://github.com/mattermost/xml-roundtrip-validator/commit/460296229913fb6653021a48599a3a0ff83802e6#diff-ed639bc2f5aff893[…]93910d34aa84797b3b8R34

Please correct me if I'm wrong.

jlourenc commented:
@beorc Thanks for the update regarding https://github.com/mattermost/xml-roundtrip-validator

@crewjam I believe I have now fixed the conflicting tests by splitting them using build flags.
I've also modified the CI so that tests are run against the different supported Go versions.

Let me know your thoughts.

beorc commented:
@crewjam Since this is a security issue and has nothing to do with failed tests could you please expedite the action in this regard?

crewjam commented:
Thanks all!
